### ISSUE TITLE
Classloader issues cause maven plugin (v1.14.251) to fail when using classesFromAutomaticJaxrsApplication

### ISSUE THREAD

**Issue by sayresVT** [vojtechhabarta/typescript-generator#142]
Using 1.14.251 of this plugin, we've been running into sporadic maven build issues that appear to be related to the classloading mechanism that's used when the classesFromAutomaticJaxrsApplication flag is set to `true`.  When running our multi-module maven build in parallel, the plugin will occasionally attempt to load a `@Path` annotated class that is from another module that isn't a direct dependency of the project running the plugin.  When it attempts to load the class, a `java.lang.ClassNotFoundException` is thrown and the build fails.

We had attempted to add the package of the class to the excludeClassPatterns, but that didn't help due to the fact that the Input/JaxrsApplicationScanner classes perform their classloading before any filtering occurs.  This is problematic because the FastClasspathScanner is occasionally finding classes that aren't available to the thread context class loader.  I was thinking there might be a couple of ways to address this:

1.  Perform class exclusion filtering on the FastClasspathScanner using the [constructor scan spec](https://github.com/lukehutch/fast-classpath-scanner/wiki/2.-Constructor#scan-spec).
2.  Perform exclusion filtering on class names found by the FastClasspathScanner before attempting to load them
3.  Allow individual classloading failures to be suppressed without causing the entire plugin execution to fail

We can update to the latest version of the plugin, but from what I can tell, the classpath scanning/classloading mechanism hasn't changed in a way that would prevent us from encountering this issue.  This is the relevant stack trace from our CI server when this issue causes the build to fail:
```
12:20:30 org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal cz.habarta.typescript-generator:typescript-generator-maven-plugin:1.14.251:generate (generate) on project PROJECT: Execution generate of goal cz.habarta.typescript-generator:typescript-generator-maven-plugin:1.14.251:generate failed: Cannot load JAX-RS application. For more information see https://github.com/vojtechhabarta/typescript-generator/wiki/JAX-RS-Application.
12:20:30 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
12:20:30 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
12:20:30 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
12:20:30 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
12:20:30 	at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:185)
12:20:30 	at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:181)
12:20:30 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
12:20:30 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
12:20:30 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
12:20:30 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
12:20:30 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
12:20:30 	at java.lang.Thread.run(Thread.java:745)
12:20:30 Caused by: org.apache.maven.plugin.PluginExecutionException: Execution generate of goal cz.habarta.typescript-generator:typescript-generator-maven-plugin:1.14.251:generate failed: Cannot load JAX-RS application. For more information see https://github.com/vojtechhabarta/typescript-generator/wiki/JAX-RS-Application.
12:20:30 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:145)
12:20:30 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
12:20:30 	... 11 more
12:20:30 Caused by: java.lang.RuntimeException: Cannot load JAX-RS application. For more information see https://github.com/vojtechhabarta/typescript-generator/wiki/JAX-RS-Application.
12:20:30 	at cz.habarta.typescript.generator.JaxrsApplicationScanner.scanJaxrsApplication(JaxrsApplicationScanner.java:32)
12:20:30 	at cz.habarta.typescript.generator.Input.fromJaxrsApplication(Input.java:63)
12:20:30 	at cz.habarta.typescript.generator.Input.fromClassNamesAndJaxrsApplication(Input.java:82)
12:20:30 	at cz.habarta.typescript.generator.maven.GenerateMojo.execute(GenerateMojo.java:331)
12:20:30 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
12:20:30 	... 12 more
12:20:30 Caused by: java.lang.ClassNotFoundException: com.package.RestAPI
12:20:30 	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
12:20:30 	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
12:20:30 	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
12:20:30 	at cz.habarta.typescript.generator.JaxrsApplicationScanner.scanClasspathForJaxrsResources(JaxrsApplicationScanner.java:50)
12:20:30 	at cz.habarta.typescript.generator.JaxrsApplicationScanner.scanJaxrsApplication(JaxrsApplicationScanner.java:24)
12:20:30 	... 16 more
```

**Comment by vojtechhabarta** [comment]
Thanks for detailed description of the problem.

You think there was no relevant change since 1.14.251 but nevertheless I would try to update typescript-generator to latest version for two reasons:
1. there was `fast-classpath-scanner` update in version 1.22.307 which could change something
2. more convenient stack traces (up-to-date line numbers)

**Comment by sayresVT** [comment]
Sounds good!  I'll update to the latest version and report back.  Unfortunately, due to the inconsistent nature of the problem, we might not be able to immediately confirm whether the latest version has addressed the issue or not.

**Comment by sayresVT** [comment]
So after attempting to upgrade to the latest version, I encountered another issue where some of our types were no longer being discovered (and therefore generated).  I narrowed it down to a change that was introduced in 1.17.282/284 after verifying that it was still working correctly in 1.16.274.  

The "issue" may have been caused by some unintended behavior that we were relying on before.  One of our JAX-RS interfaces doesn't contain any methods, but instead, it extends a generic interface that includes the necessary JAX-RS methods. In version 1.16.274, our JAX-RS interface was being parsed by the `Jackson2Parser.parseBean()` method.  It appears that the Jackson2Parser was capturing the parent interface type parameter information in [this section](https://github.com/vojtechhabarta/typescript-generator/blob/f31f2d3aa29fbc6224df6e521eedcd74ee8c12ca/typescript-generator-core/src/main/java/cz/habarta/typescript/generator/parser/Jackson2Parser.java#L120-L122).  However, in version 1.17.284 & 1.25.322, the `JaxrsApplicationParser.tryParse()` method ends up processing the interface (which skips the `parseBean()` method) without picking up the additional type information included in the parent interfaces.  This prevents some additional types from ever being discovered.

I'm thinking that I might be able to work around this by explicitly listing the additional types on the maven configuration, but should I go ahead and create a separate issue for this in the meantime?

**Comment by vojtechhabarta** [comment]
You are right. Version 1.17 added support for generating interface for JAX-RS application which was bigger change. JAX-RS resource classes are not processed by `Jackson2Parser` now. This probably was incompatible change in some cases but I think it was just incorrect to process JAX-RS application class and resource classes using `Jackson2Parser`. These classes are now processed using `JaxrsApplicationParser`.

JAX-RS is relatively complex and typescript-generator definitely doesn't behave exactly as JAX-RS specifies. So it would be great if you could fill an issue with example what typescript-generator should handle better. But in general it is not always possible to detect all used classes from _statically_ typed resources which can behave _dynamically_. On the other hand in many cases we can use annotations like `@JsonSubTypes` from Jackson 2 or `@ApiOperation` from Swagger to specify additional types.

If some class is not detected automatically you can add it using `classes` or `classPatterns` parameter.

**Comment by sayresVT** [comment]
I was able to verify that the classloader issues are still occurring in version 1.25.322.  This is the latest stack trace from our CI server:
```
18:09:39 org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal cz.habarta.typescript-generator:typescript-generator-maven-plugin:1.25.322:generate (generate) on project PROJECT: Execution generate of goal cz.habarta.typescript-generator:typescript-generator-maven-plugin:1.25.322:generate failed: Cannot load JAX-RS application. For more information see https://github.com/vojtechhabarta/typescript-generator/wiki/JAX-RS-Application.
18:09:39 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
18:09:39 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
18:09:39 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
18:09:39 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
18:09:39 	at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:185)
18:09:39 	at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:181)
18:09:39 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
18:09:39 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
18:09:39 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
18:09:39 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
18:09:39 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
18:09:39 	at java.lang.Thread.run(Thread.java:745)
18:09:39 Caused by: org.apache.maven.plugin.PluginExecutionException: Execution generate of goal cz.habarta.typescript-generator:typescript-generator-maven-plugin:1.25.322:generate failed: Cannot load JAX-RS application. For more information see https://github.com/vojtechhabarta/typescript-generator/wiki/JAX-RS-Application.
18:09:39 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:145)
18:09:39 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
18:09:39 	... 11 more
18:09:39 Caused by: java.lang.RuntimeException: Cannot load JAX-RS application. For more information see https://github.com/vojtechhabarta/typescript-generator/wiki/JAX-RS-Application.
18:09:39 	at cz.habarta.typescript.generator.JaxrsApplicationScanner.reportError(JaxrsApplicationScanner.java:51)
18:09:39 	at cz.habarta.typescript.generator.JaxrsApplicationScanner.scanAutomaticJaxrsApplication(JaxrsApplicationScanner.java:43)
18:09:39 	at cz.habarta.typescript.generator.Input.fromClassNamesAndJaxrsApplication(Input.java:50)
18:09:39 	at cz.habarta.typescript.generator.maven.GenerateMojo.execute(GenerateMojo.java:477)
18:09:39 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
18:09:39 	... 12 more
18:09:39 Caused by: java.lang.ClassNotFoundException: com.package.RestAPI
18:09:39 	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
18:09:39 	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
18:09:39 	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
18:09:39 	at cz.habarta.typescript.generator.JaxrsApplicationScanner.scanAutomaticJaxrsApplication(JaxrsApplicationScanner.java:38)
18:09:39 	... 15 more
```

**Comment by vojtechhabarta** [comment]
Is problematic class (`com.package.RestAPI`) from some dependency or is it directly from the module where typescript-generator is run? Ideally if you could send me your pom.xml file. Or please check at least if you have typescript-generator in dependencies (should not) and also if jar containing problematic class is not in typescript-generator dependencies. (Maybe run maven with `-X` switch to see detailed information how `typescript-generator-maven-plugin` is invoked.)

I would really like to know why this can happen before making some changes.

**Comment by sayresVT** [comment]
The problematic class is from another one of our modules in our multi-module maven project.  However, the module is not listed as an explicit dependency in the pom with the typescript-generator plugin.  Also, running the dependency:tree goal does not list the module as a transitive dependency so we're not entirely sure how it's even being discovered.

To answer your other question, the typescript-generator is not defined as a dependency in our pom (only as a plugin).  Also, the typescript-generator plugin definition in our pom does not include any direct dependencies.

I ran maven with the -X flag just on the individual project containing the typescript-generator plugin, but it seem to indicate anything related to the problematic class.  I'll try running the full build with the -X flag locally and possibly on the CI server to see if I can find anything that way.

**Comment by sayresVT** [comment]
From what I can tell so far, the issue only seems to occur whenever the maven build is running on the module with the problematic class at the same time as the module with the typescript-generator plugin.  During the successful builds on our CI server, maven has already finished running on the module with the problematic class before the typescript-generator plugin begins execution.

**Comment by vojtechhabarta** [comment]
I have another question: are you invoking typescript-generator directly using `mvn typescript-generator:generate` or do you run it as part of some maven phase like `mvn install`?

And how do you parallelize the build? I guess the problem is somehow related to this parallelization, right?

**Comment by sayresVT** [comment]
No, we're not running the typescript-generator directly.  We're running `mvn clean install -T 12` on the top level parent project of our multi-module project (although the clean probably isn't necessary since we're starting from a fresh workspace every time).  

Just in case you're unfamiliar with the option, the `-T 12` is telling maven to attempt to use up to 12 threads to build individual modules [in parallel](https://cwiki.apache.org/confluence/display/MAVEN/Parallel+builds+in+Maven+3) while still respecting the reactor order.  I would say the issue is definitely somewhat related to the parallelization because we've never encountered this issue before with our non-parallel production builds.

**Comment by vojtechhabarta** [comment]
Since we sometimes get classes from fast-classpath-scanner that are not part of current module I think right solution would be to correctly configure fast-classpath-scanner so that we only get relevant classes. I asked an question how to do it in fast-classpath-scanner project.

If it will not work I would implement some mitigation of the problem like you proposed. My intent is to not force users to configure exclusion for classes that should not even be there so I would probably implement your 3rd point.

**Comment by lukehutch** [comment]
I just released v2.2.0 of FastClasspathScanner. It may solve the issue you are seeing with `mvn clean install -T 12` -- please let me know. If not, please test the other versions of FCS detailed in my comment on the bug report created by @vojtechhabarta, linked above. Thanks!

**Comment by lukehutch** [comment]
v2.2.0 didn't make it to Maven Central due to a Maven .pom issue, but v2.2.1 is there. Please let me know if this version solves the parallel build issue. 

PS you really should use `Class<?> ScanResult#classNameToClassRef(String className)` or `List<Class<?>> ScanResult#classNamesToClassRefs(List<String> classNames)` to load classes, given a `ScanResult` returned by FastClasspathScanner, because this ensures that the correct `ClassLoader` for a given classpath element is used to load all classes found in that classpath element. (In other words, don't use `Class.forName()`, and in general don't use `ClassLoader.loadClass()`, unless you're absolutely sure you have the right `ClassLoader`.)

**Comment by vojtechhabarta** [comment]
@sayresVT I made some improvements in classpath scanning.

Most importantly scanning is now limited to project transitive dependencies. This, I hope, could resolve the problem.

As a backup, if class fails to load the problem is logged and processing continues. You should find `Error: Cannot load class ...` message in log. In this case you can switch verbose output on using `<debug>true</debug>` parameter to see where problematic class came from.

Changes are in [1.26.333](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v1.26.333) release. Please try it and let me know.

**Comment by lukehutch** [comment]
@vojtechhabarta if you load classes the way I described in the previous comment, the class not found errors should never happen.

I'm still interested to know if the parallel Maven build issue was resolved with FastClasspathScanner v2.2.1.

**Comment by sayresVT** [comment]
@vojtechhabarta thanks for making those changes!  I noticed that it was released this morning and immediately started working on rolling out the update to v1.26.333.

Unfortunately, changes for #149 appear to have caused us a few build errors.  However, I'm thinking the errors are probably related to some quirky class/interface inheritance going on in our application so I'm seeing if they can be resolved with a little restructuring and a more logical inheritance hierarchy.  I should hopefully have it running on all of our CI builds very soon.

**Comment by sayresVT** [comment]
@vojtechhabarta the update to the latest version of the plugin was merged into our master last night and so far the issue hasn't been seen again.  I also searched through the logs and so far haven't seen the statement for `Cannot load class`.

**Comment by sayresVT** [comment]
Closing out the issue as it hasn't been encountered in over a week after upgrading to v1.26.333

**Comment by vojtechhabarta** [comment]
@sayresVT great, thanks for the feedback.
